src/main/java/com/epam/aidial/core/limiter/RateLimiter.java

```diff
@@ -81,22 +85,36 @@ public RateLimitResult limit(ProxyContext context) {
     public boolean register(ProxyContext context) {
         String traceId = context.getTraceId();
         Entity entity = traceIdToEntity.get(traceId);
-        if (entity != null) {
+        boolean result = entity != null;
+        if (result) {
+            entity.link(context.getParentSpanId(), context.getCurrentSpanId());
             // update context with the original requester
-            if (entity.user()) {
-                context.setUserHash(entity.name());
+            if (entity.isUser()) {
+                context.setUserHash(entity.getName());
             } else {
-                context.setOriginalProject(entity.name());
+                context.setOriginalProject(entity.getName());
             }
         } else {
             if (context.getKey() != null) {
                 Key key = context.getKey();
-                traceIdToEntity.put(traceId, new Entity(key.getKey(), Collections.singletonList(key.getRole()), key.getProject(), false));
+                entity = new Entity(key.getKey(), Collections.singletonList(key.getRole()), key.getProject(), false);
+                traceIdToEntity.put(traceId, entity);
             } else {
-                traceIdToEntity.put(traceId, new Entity(context.getUserSub(), context.getUserRoles(), context.getUserHash(), true));
+                entity = new Entity(context.getUserSub(), context.getUserRoles(), context.getUserHash(), true);
+                traceIdToEntity.put(traceId, entity);
             }
         }
-        return entity != null;
+        entity.addCall(context.getCurrentSpanId());
+        return result;
+    }
+
+    public void calculateTokenUsage(ProxyContext context) {
+        Entity entity = getEntityFromTracingContext(context);
+        if (entity == null) {
+            log.warn("Entity is not found by traceId={}", context.getTraceId());
```

Comment:
Please add prefix to locate the problem more easily: "Failed to calculate token usage"